### ISSUE TITLE
CopperSpice 1.9.1 with clang and libc++ on Linux: segmentation fault when terminating QFileDialog with Esc key

### ISSUE THREAD

**Issue by e-mattes** [copperspice/copperspice#295]
As the rest of our software is built with -stdlib=libc++ on Linux, I want/need to build CopperSpice with -stdlib=libc++, too.
Everything works but for one problem: when a QFileDialog is terminated with the Esc key, I get a segmentation fault.

This does not happen if I terminate the QFileDialog by any other means (though if I use any keyboard-based method, I get a BadWindow XCB Error).

This also does not happen if I omit -stdlib=libc++ when building CopperSpice and the test program.
It also does not happen on macOS, therefore the problem may be related to CsGuiXcb1.9.so.

I tried clang++ 15 and 18, doesn't make a difference.

Here's the top of the stacktrace:

Thread 1 "FileOpen" received signal SIGSEGV, Segmentation fault.
0x00007ffff63e2812 in QCompleter::eventFilter (this=0x555555bccfc0, 
    o=0x555555e5f680, e=0x7fffffff77e8)
    at /u/src/copperspice-cs-1.9.1/src/gui/util/qcompleter.cpp:1224
1224       if (d->eatFocusOut && o == d->widget && e->type() == QEvent::FocusOut) {
(gdb) p d
$1 = (QCompleterPrivate * const) 0x0
(gdb) bt
#0  0x00007ffff63e2812 in QCompleter::eventFilter (this=0x555555bccfc0, 
    o=0x555555e5f680, e=0x7fffffff77e8)
    at /u/src/copperspice-cs-1.9.1/src/gui/util/qcompleter.cpp:1224
#1  0x00007ffff7b15d8e in QCoreApplicationPrivate::sendThroughObjectEventFilters (this=0x555555588330, receiver=0x555555e5f680, event=0x7fffffff77e8)
    at /u/src/copperspice-cs-1.9.1/src/core/kernel/qcoreapplication.cpp:797

Here's the test program, the script I use for building Copperspice, and the complete stacktrace:
[FileOpen.zip](https://github.com/user-attachments/files/16597165/FileOpen.zip)

**Comment by e-mattes** [comment]
Interesting observation: In this test program (which overrides keyPressEvent), pressing Esc causes clicking the Cancel button to crash the application, that is, having received that QKeyEvent causes the destructor to crash even if that event is ignored by the dialog.
[FileOpen2.zip](https://github.com/user-attachments/files/16598252/FileOpen2.zip)

**Comment by e-mattes** [comment]
It's not only the Esc key! For instance, if I press Key_Up and then click on the Cancel button, the application crashes in the same way. Ditto when I enter something in the "File name" field. Key_Tab doesn't have that effect.

**Comment by e-mattes** [comment]
valgrind doesn't report any issues until QCompleter::eventFilter() tries to reference that null pointer.

**Comment by agserm** [comment]
Thanks for the very thorough report! We will investigate this. Can you let us know what distribution of Linux you are using?

**Comment by e-mattes** [comment]
I tried Debian 10, Debian 11, Debian 12, Gentoo. Same behavior on all of them.

**Comment by e-mattes** [comment]
QPrintDialog is also affected. Esc doesn't cause trouble here, but changing the output file name with the keyboard and then clicking on Cancel causes a segmentation fault.

**Comment by e-mattes** [comment]
I couldn't recreate the problem on macOS with the test program attached above, but a bigger program does crash when hitting Esc when a QFileDialog is displayed. The stack looks like it does on Linux:

Process 5849 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x5d)
    frame #0: 0x0000000110c7f92c libCsGui1.9.dylib`QCompleter::eventFilter(QObject*, QEvent*) + 56
libCsGui1.9.dylib`QCompleter::eventFilter:
->  0x110c7f92c <+56>: ldrb   w8, [x8, #0x5d]
    0x110c7f930 <+60>: tbz    w8, #0x0, 0x110c7f9c4     ; <+208>
    0x110c7f934 <+64>: b      0x110c7f938               ; <+68>
    0x110c7f938 <+68>: ldur   x8, [x29, #-0x28]
Target 0: (myapp) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x5d)
  * frame #0: 0x0000000110c7f92c libCsGui1.9.dylib`QCompleter::eventFilter(QObject*, QEvent*) + 56
    frame #1: 0x000000010698714c libCsCore1.9.dylib`QCoreApplicationPrivate::sendThroughObjectEventFilters(QObject*, QEvent*) + 248
    frame #2: 0x000000010fe07a38 libCsGui1.9.dylib`QApplicationPrivate::notify_helper(QObject*, QEvent*) + 424
    frame #3: 0x000000010fe0ca80 libCsGui1.9.dylib`QApplication::notify(QObject*, QEvent*) + 13116
    frame #4: 0x0000000106986a78 libCsCore1.9.dylib`QCoreApplication::notifyInternal(QObject*, QEvent*) + 88
    frame #5: 0x000000010f490dcc libCsGui1.9.dylib`QCoreApplication::sendEvent(QObject*, QEvent*) + 120
    frame #6: 0x00000001069cb440 libCsCore1.9.dylib`QObject::cs_forceRemoveChild() + 116
    frame #7: 0x0000000110d43de0 libCsGui1.9.dylib`QWidget::~QWidget() + 760
    frame #8: 0x0000000110d45098 libCsGui1.9.dylib`QWidget::~QWidget() + 36
    frame #9: 0x0000000110d45120 libCsGui1.9.dylib`QWidget::~QWidget() + 28
    frame #10: 0x00000001069c60e8 libCsCore1.9.dylib`QObject::deleteChildren() + 204
    frame #11: 0x0000000106a86b70 libCsCore1.9.dylib`CSInternalChildren::deleteChildren(QObject*) + 48
    frame #12: 0x0000000110d44370 libCsGui1.9.dylib`QWidget::~QWidget() + 2184
    frame #13: 0x0000000110e2e07c libCsGui1.9.dylib`QFrame::~QFrame() + 40
    frame #14: 0x0000000110d348ac libCsGui1.9.dylib`QAbstractScrollArea::~QAbstractScrollArea() + 144
    frame #15: 0x000000010fabbaa0 libCsGui1.9.dylib`QAbstractItemView::~QAbstractItemView() + 240
    frame #16: 0x000000010fc06ce0 libCsGui1.9.dylib`QListView::~QListView() + 40
    frame #17: 0x000000010fc06d14 libCsGui1.9.dylib`QListView::~QListView() + 36
    frame #18: 0x000000010fc06d9c libCsGui1.9.dylib`QListView::~QListView() + 28
    frame #19: 0x0000000110c8eb68 libCsGui1.9.dylib`QCompleterPrivate::~QCompleterPrivate() + 88
    frame #20: 0x0000000110c8138c libCsGui1.9.dylib`QCompleterPrivate::~QCompleterPrivate() + 28
    frame #21: 0x0000000110c813b8 libCsGui1.9.dylib`QCompleterPrivate::~QCompleterPrivate() + 28
    frame #22: 0x0000000110c83e68 libCsGui1.9.dylib`std::__1::default_delete<QCompleterPrivate>::operator()[abi:v160006](QCompleterPrivate*) const + 60
    frame #23: 0x0000000110c83dd4 libCsGui1.9.dylib`std::__1::unique_ptr<QCompleterPrivate, std::__1::default_delete<QCompleterPrivate>>::reset[abi:v160006](QCompleterPrivate*) + 104
    frame #24: 0x0000000110c83d5c libCsGui1.9.dylib`std::__1::unique_ptr<QCompleterPrivate, std::__1::default_delete<QCompleterPrivate>>::~unique_ptr[abi:v160006]() + 32
    frame #25: 0x0000000110c83d2c libCsGui1.9.dylib`std::__1::unique_ptr<QCompleterPrivate, std::__1::default_delete<QCompleterPrivate>>::~unique_ptr[abi:v160006]() + 28
    frame #26: 0x0000000110c83d00 libCsGui1.9.dylib`CsPointer::CsUniquePointer<QCompleterPrivate, std::__1::default_delete<QCompleterPrivate>>::~CsUniquePointer() + 28
    frame #27: 0x0000000110c83cd4 libCsGui1.9.dylib`QUniquePointer<QCompleterPrivate, std::__1::default_delete<QCompleterPrivate>>::~QUniquePointer() + 28
    frame #28: 0x0000000110c83ca8 libCsGui1.9.dylib`QScopedPointer<QCompleterPrivate, std::__1::default_delete<QCompleterPrivate>>::~QScopedPointer() + 28
    frame #29: 0x0000000110c7d2c0 libCsGui1.9.dylib`QScopedPointer<QCompleterPrivate, std::__1::default_delete<QCompleterPrivate>>::~QScopedPointer() + 28
    frame #30: 0x0000000110c7db98 libCsGui1.9.dylib`QCompleter::~QCompleter() + 124
    frame #31: 0x000000010f4e65c8 libCsGui1.9.dylib`QFSCompleter::~QFSCompleter() + 40
    frame #32: 0x000000010f4ce62c libCsGui1.9.dylib`QFSCompleter::~QFSCompleter() + 36
    frame #33: 0x000000010f4ce670 libCsGui1.9.dylib`QFSCompleter::~QFSCompleter() + 28
    frame #34: 0x00000001069c60e8 libCsCore1.9.dylib`QObject::deleteChildren() + 204
    frame #35: 0x0000000106a86b70 libCsCore1.9.dylib`CSInternalChildren::deleteChildren(QObject*) + 48
    frame #36: 0x0000000110d44370 libCsGui1.9.dylib`QWidget::~QWidget() + 2184
    frame #37: 0x000000010f48f72c libCsGui1.9.dylib`QDialog::~QDialog() + 164
    frame #38: 0x000000010f4c28b4 libCsGui1.9.dylib`QFileDialog::~QFileDialog() + 140

Again, CopperSpice and app built with -stdlib=libc++.

**Comment by e-mattes** [comment]
What evil will happen if I remove the QObject::cs_forceRemoveChild() call? The git history doesn't reach back far enough...

**Comment by e-mattes** [comment]
That bigger program also crashes on Windows when I try to open(!) a QFileDialog. CopperSpice and app built with Visual C++ (toolchain v143). Other dialogs work just fine.

[winstack.zip](https://github.com/user-attachments/files/16636481/winstack.zip)

**Comment by e-mattes** [comment]
Removing the QObject::cs_forceRemoveChild() call doesn't improve the situation:
[Uploading linstack.zip…]()

**Comment by e-mattes** [comment]
The crash on Windows also happens with the DLLs in https://download.copperspice.com/copperspice/binary/cs-1.9/setup-CopperSpice-1.9.1-x64-MSVC-release.exe.

**Comment by e-mattes** [comment]
The crash on Windows is caused by setOption (DontUseNativeDialog).

**Comment by e-mattes** [comment]
Here's a test program which demonstrates the crash on Windows. 
[FileOpen4.zip](https://github.com/user-attachments/files/16659334/FileOpen4.zip)

**Comment by e-mattes** [comment]
The upload of linstack.txt above didn't work. Let me try again:
[linstack.zip](https://github.com/user-attachments/files/16659554/linstack.zip)

**Comment by e-mattes** [comment]
Here's the call stack for the DontUseNativeDialog crash on Windows (FileOpen4.zip):
[winstack2.zip](https://github.com/user-attachments/files/16671165/winstack2.zip)

**Comment by e-mattes** [comment]
Building CopperSpice with QT_NO_COMPLETER works around the crash on Linux.

**Comment by e-mattes** [comment]
This patch makes the DontUseNativeDialog crash on Windows go away. However, it may create a memory leak.
[DontUseNativeDialog-1.zip](https://github.com/user-attachments/files/16672117/DontUseNativeDialog-1.zip)

**Comment by e-mattes** [comment]
This patch makes the crash on Linux go away. Probably it doesn't fix the root cause.
[qcompleter-1.zip](https://github.com/user-attachments/files/16691292/qcompleter-1.zip)

**Comment by e-mattes** [comment]
My DontUseNativeDialog-1.diff patch above has an undesirable side effect: I get multiple Cancel and Open buttons.

**Comment by e-mattes** [comment]
This patch for the crash on Windows seems to be better:
[DontUseNativeDialog-2.zip](https://github.com/user-attachments/files/16711324/DontUseNativeDialog-2.zip)

**Comment by bgeller** [comment]
You have done a lot of good testing and we will review everything next week. Please let us know if you discover anything else. 

Barbara
